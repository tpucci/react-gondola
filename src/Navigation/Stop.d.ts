import { ComponentType } from 'react';

export interface IStop {
  Component: ComponentType<IStopComponentProps>;
  isFullScreen?: boolean;
  name: string;
  isAuthorized: boolean;
}

export interface IStopComponentProps {
  isAuthorized: boolean;
}
