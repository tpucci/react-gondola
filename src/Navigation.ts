import { Subject, Observable } from 'rxjs';
import { ComponentType } from 'react';
import { mergeAll, map, scan } from 'rxjs/operators';

export interface IStop {
  name: string;
  Component: ComponentType;
  isFullScreen?: boolean;
}

type Stack = IStop[];

interface IFullScreenStackProperties {
  canalId: string;
  fullScreenStack: Stack;
}

interface IFullScreenStackMap {
  [canalId: string]: Stack;
}

export class Navigation {
  static getInstance() {
    if (Navigation.instance) {
      return Navigation.instance;
    }
    Navigation.instance = new Navigation();
    return Navigation.instance;
  }
  private static instance: Navigation;

  canalsFullScreenStackProperties$ = new Subject<
    Observable<IFullScreenStackProperties>
  >();

  fullSceenStack$: Observable<
    Stack
  > = this.canalsFullScreenStackProperties$.pipe(
    mergeAll(),
    scan(
      (
        fullScreenStackMap: IFullScreenStackMap,
        fullScreenStackProperties: IFullScreenStackProperties
      ) => ({
        ...fullScreenStackMap,
        [fullScreenStackProperties.canalId]:
          fullScreenStackProperties.fullScreenStack
      }),
      {}
    ),
    map(fullScreenStackMap =>
      Object.keys(fullScreenStackMap).reduce(
        (fullScreenStack: Stack, canalId: string) =>
          fullScreenStack.concat(fullScreenStackMap[canalId]),
        []
      )
    )
  );
}
