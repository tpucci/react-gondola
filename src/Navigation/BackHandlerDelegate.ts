import { BackHandler } from 'react-native';
import { fromEventPattern, Observable, Subject } from 'rxjs';
import { share, map, merge, tap } from 'rxjs/operators';

export interface BackEvent {
  target: string | null;
}

export class BackHandlerDelegate {
  /**
   * Attach the given handler to the event source. When the source emits,
   * calls the handler. The handler is responsible to register a callback.
   * If it succeeded to register a callback, we call it.
   * If not, and no other react native back event listener has been registered
   * elsewhere, this listener returns false which might kill the app.
   * See https://facebook.github.io/react-native/docs/backhandler.html.
   * @param handler
   */
  private addEventListener = (handler: () => void) => {
    BackHandler.addEventListener('hardwareBackPress', () => {
      handler();
      return true;
    });
  };

  private softwareBack$ = new Subject<undefined>();

  private back$: Observable<BackEvent> = fromEventPattern<undefined>(this.addEventListener).pipe(
    merge(this.softwareBack$),
    tap(() => {
      // After the event has spread, we fire onBacks callbacks.
      setTimeout(() => {
        if (this.onBackCallback) {
          this.onBackCallback();
          this.onBackCallback = undefined;
        } else {
          BackHandler.exitApp();
        }
      }, 0);
    }),
    map(() => ({
      target: null,
    })),
    share()
  );

  private onBackCallback?: () => any;
  setOnBackCallback = (cb: () => any) => {
    this.onBackCallback = cb;
  };

  back = () => {
    this.softwareBack$.next();
  };

  defaultBackContext = { back$: this.back$ };
}
