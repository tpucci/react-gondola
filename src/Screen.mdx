---
name: Screen
menu: API
route: /api/screen
---

import { Props } from 'docz';
import { Screen } from 'react-nonav';

# Screen

`type: React.Component`

## Props

|    Prop name    | Required | Default     | Usage                                                                                                                                                                                     |
| :-------------: | :------: | ----------- | ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `isFullScreen`  | `false`  | `false`     | Specifies if the screen should be render in the `FullScreenPortal`.                                                                                                                       |
|    `onBack`     | `false`  | `undefined` | Function to call when a back event occurs. Use this prop and `visible` to control your Canal.                                                                                             |
|     `name`      |  `true`  | `undefined` | Unique identifier of the screen.                                                                                                                                                          |
|   `Component`   |  `true`  | `undefined` | Component to render. This is your page.                                                                                                                                                   |
|     `props`     | `false`  | `undefined` | Props to pass to the Component. Those props are transition-safe. This means while a transition occurs, those props are frozen. When the transition completes, the props flow is released. |
| `Transtitioner` | `false`  | `None`      | Transition to apply to the screen when it appears or disappears.                                                                                                                          |
|    `visible`    |  `true`  | `false`     | Visibility of your screen. Use this prop and `onBack` to control your Canal.                                                                                                              |
