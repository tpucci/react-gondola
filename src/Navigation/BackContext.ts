import { createContext } from 'react';
import { Observable } from 'rxjs';
import { IBackEvent } from './BackHandlerDelegate';
import { Navigation } from './Navigation';

interface IBackContext {
  back$: Observable<IBackEvent>;
}

export const BackContext = createContext<IBackContext>(
  Navigation.instance.backHandlerDelegate.defaultBackContext
);
