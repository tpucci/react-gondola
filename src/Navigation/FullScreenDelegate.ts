import { Subject, Observable } from 'rxjs';
import { map, scan, tap } from 'rxjs/operators';
import { ReactElement } from 'react';
import { ScreenProps, Screen } from '../Screen';

type Stack = ReactElement<ScreenProps, typeof Screen>[];

interface FullScreenStackProperties {
  canalId: string;
  fullScreenStack: Stack;
}

interface FullScreenStackMap {
  [canalId: string]: Stack;
}

export class FullScreenDelegate {
  canalsFullScreenStackProperties$ = new Subject<FullScreenStackProperties>();

  fullSceenStack$: Observable<Stack> = this.canalsFullScreenStackProperties$.pipe(
    scan(
      (
        fullScreenStackMap: FullScreenStackMap,
        fullScreenStackProperties: FullScreenStackProperties
      ) => ({
        ...fullScreenStackMap,
        [fullScreenStackProperties.canalId]: fullScreenStackProperties.fullScreenStack,
      }),
      {}
    ),
    map((fullScreenStackMap: FullScreenStackMap) =>
      Object.keys(fullScreenStackMap).reduce(
        (fullScreenStack: Stack, canalId: string) =>
          fullScreenStack.concat(fullScreenStackMap[canalId]),
        []
      )
    ),
    tap(o => console.warn(o))
  );
}
