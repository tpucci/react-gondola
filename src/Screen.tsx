import React, { Component as ReactComponent, createFactory, ComponentType } from 'react';
import { BackContext } from './Navigation/BackContext';
import { filter, share, tap } from 'rxjs/operators';
import { Navigation } from './Navigation';
import { TransitionComponentType } from './transitions/Transition';
import { None } from './transitions';
import { withBackContext, WithBackContext } from './withBackContext';

export interface ScreenProps {
  isFullScreen?: boolean;
  onBack?: (() => any) | undefined;
  name: string;
  Component: ComponentType<any>;
  props?: object | undefined;
  Transitioner?: TransitionComponentType;
  visible: boolean;
}

class ScreenComponent extends ReactComponent<WithBackContext<ScreenProps>, ScreenProps['props']> {
  constructor(props: WithBackContext<ScreenProps>) {
    super(props);
    this.state = props.props;
  }

  state: ScreenProps['props'];
  static getDerivedStateFromProps(
    props: WithBackContext<ScreenProps>,
    state: ScreenProps['props']
  ) {
    if (!props.visible) return state;
    return props.props;
  }

  static defaultProps = {
    isFullScreen: false,
    Transitioner: None,
    props: {},
  };

  back$ = this.props.backContext.back$.pipe(
    filter(value => value.target === this.props.name),
    tap(() => {
      if (this.props.onBack) {
        Navigation.instance.backHandlerDelegate.setOnBackCallback(this.props.onBack);
      }
    }),
    share()
  );

  backSubscription = this.back$.subscribe();

  componentWillUnmount() {
    this.backSubscription.unsubscribe();
  }

  /**
   * @TODO 2019-07-26 Update @types/react once https://github.com/DefinitelyTyped/DefinitelyTyped/pull is merged.
   */
  // @ts-ignore
  factory = createFactory(this.props.Component);

  render() {
    const { Transitioner, onBack } = this.props;
    return (
      /**
       *  @todo Transitioner is always defined in static defaultProps.
       */
      // @ts-ignore
      <Transitioner directionForward={this.props.visible}>
        <BackContext.Provider
          value={{
            back$: this.back$,
          }}
        >
          {this.factory({ navigation: { goBack: onBack }, ...this.state })}
        </BackContext.Provider>
      </Transitioner>
    );
  }
}

// @ts-ignore
export const Screen = withBackContext(ScreenComponent);
