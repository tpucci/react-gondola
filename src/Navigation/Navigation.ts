import { FullScreenDelegate } from './FullScreenDelegate';
import { BackHandlerDelegate } from './BackHandlerDelegate';

export class Navigation {
  static get instance() {
    if (Navigation.singleton) {
      return Navigation.singleton;
    }
    Navigation.singleton = new Navigation();
    return Navigation.singleton;
  }

  private constructor() {}
  private static singleton: Navigation;

  backHandlerDelegate = new BackHandlerDelegate();

  fullScreenDelegate = new FullScreenDelegate();

  back = () => {
    this.backHandlerDelegate.back();
  };
}
