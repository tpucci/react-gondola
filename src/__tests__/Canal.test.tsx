import React from 'react';
import { Text, View } from 'react-native';
import TestRenderer from 'react-test-renderer';
import { Canal } from '../Canal';
import { Screen } from '../Screen';
import { Subject } from 'rxjs';
import { BackEvent } from '../Navigation/BackHandlerDelegate';
import { BackContext } from '../Navigation/BackContext';
import { Navigation } from '../Navigation';

describe('Canal', () => {
  it('renders its children', () => {
    const testRenderer = TestRenderer.create(
      <Canal>
        <Screen Component={() => <Text>a</Text>} name="a" visible />
        <Screen Component={() => <Text>b</Text>} name="b" visible />
      </Canal>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
  it('renders its only child', () => {
    const testRenderer = TestRenderer.create(
      <Canal>
        <Screen Component={() => <Text>a</Text>} name="a" visible />
      </Canal>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
  it('renders only non-fullscreen children', () => {
    const testRenderer = TestRenderer.create(
      <Canal>
        <Screen Component={() => <Text>a</Text>} name="a" visible />
        <Screen Component={() => <Text>b</Text>} name="b" visible isFullScreen />
      </Canal>
    );
    expect(testRenderer.toJSON()).toMatchSnapshot();
  });
  it('notifies the fullScreenDelegate of the full-screen screens after any render', () => {
    const spy = jest.spyOn(
      Navigation.instance.fullScreenDelegate.canalsFullScreenStackProperties$,
      'next'
    );
    const ScreenB = <Screen Component={() => <Text>b</Text>} name="b" visible isFullScreen />;
    const testRenderer = TestRenderer.create(
      <Canal>
        <Screen Component={() => <Text>a</Text>} name="a" visible />
        {ScreenB}
      </Canal>
    );
    expect(spy).toHaveBeenCalledWith({ canalId: '0', fullScreenStack: [ScreenB] });
    const ScreenC = (
      <Screen Component={() => <Text>c</Text>} name="c" visible={false} isFullScreen />
    );
    testRenderer.update(
      <Canal>
        <Screen Component={() => <Text>a</Text>} name="a" visible />
        {ScreenC}
      </Canal>
    );
    expect(spy).toHaveBeenCalledWith({ canalId: '0', fullScreenStack: [ScreenC] });
  });
  it('notifies the fullScreenDelegate on unmount', () => {
    const spy = jest.spyOn(
      Navigation.instance.fullScreenDelegate.canalsFullScreenStackProperties$,
      'next'
    );
    const ScreenB = <Screen Component={() => <Text>b</Text>} name="b" visible isFullScreen />;
    const testRenderer = TestRenderer.create(
      <Canal>
        <Screen Component={() => <Text>a</Text>} name="a" visible />
        {ScreenB}
      </Canal>
    );
    expect(spy).not.toHaveBeenCalledWith({ canalId: '0', fullScreenStack: [] });
    expect(spy).toHaveBeenCalledWith({ canalId: '0', fullScreenStack: [ScreenB] });
    testRenderer.update(<View></View>);
    expect(spy).toHaveBeenCalledWith({ canalId: '0', fullScreenStack: [] });
  });
  it('passes back events on', () => {
    const back$ = new Subject<BackEvent>();
    const spy = jest.fn();
    const testRenderer = TestRenderer.create(
      <BackContext.Provider value={{ back$ }}>
        <Canal>
          <Screen Component={() => <Text>a</Text>} name="a" visible={true} />
          <Screen Component={() => <Text>b</Text>} name="b" visible={false} />
        </Canal>
      </BackContext.Provider>
    );
    // @ts-ignore
    testRenderer.root.children[0].instance.back$.subscribe(spy);
    back$.next({ target: null });
    expect(spy).toHaveBeenCalledWith({
      target: 'a',
    });
    testRenderer.update(
      <BackContext.Provider value={{ back$ }}>
        <Canal>
          <Screen Component={() => <Text>a</Text>} name="a" visible={true} />
          <Screen Component={() => <Text>b</Text>} name="b" visible={true} />
        </Canal>
      </BackContext.Provider>
    );
    back$.next({ target: null });
    expect(spy).toHaveBeenCalledWith({
      target: 'b',
    });
    testRenderer.update(
      <BackContext.Provider value={{ back$ }}>
        <Canal>
          <Screen Component={() => <Text>a</Text>} name="a" visible={false} />
          <Screen Component={() => <Text>b</Text>} name="b" visible={false} />
        </Canal>
      </BackContext.Provider>
    );
    back$.next({ target: null });
    expect(spy).toHaveBeenCalledWith({
      target: null,
    });
  });
});
