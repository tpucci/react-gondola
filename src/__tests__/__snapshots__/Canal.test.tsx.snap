// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Canal renders its children 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <Text>
      a
    </Text>
  </View>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <Text>
      b
    </Text>
  </View>
</View>
`;

exports[`Canal renders its only child 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <Text>
      a
    </Text>
  </View>
</View>
`;

exports[`Canal renders only non-fullscreen children 1`] = `
<View
  style={
    Object {
      "bottom": 0,
      "left": 0,
      "position": "absolute",
      "right": 0,
      "top": 0,
    }
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <Text>
      a
    </Text>
  </View>
</View>
`;
